### PR TITLE
2026 updates

### PR DESCRIPTION
**Pull request by TeunHuijben**
Main changes: 

- split task 6 into two subtasks (1. helper function, 2. full U-Net)
- add answers to all questions
- typos/phrasing 
- add cropped images per component task
- add some links/references to specific nn.modules

Authors: 
- @TeunHuijben 
- @imancebola

### CODE REVIEW

File: solution.py [hunk 1]
```diff
@@ -410,9 +415,22 @@ def forward(self, x):
...
+#
+# ```
+# if padding == "same": `
+#     pad = kernel_size // 2
+# elif padding == "valid":
+#     pad = 0
+# else:
+#     raise ValueError("Invalid padding mode")
+# ```
+
```
Comment (AlbertDominguez) [inline solution.py:424]:
```suggestion
```

I think this should be removed?

File: solution.py [hunk 2]
```diff
@@ -726,11 +857,45 @@ def compute_fmaps_decoder(self, level: int) -> tuple[int, int]:
...
+# - Build the `Downsample`, `Upsample`, `CropAndConcat`, and `OutputConv` modules. **Reuse the classes you already wrote** in Part 1 — you do not need to re-implement them. Use `torch.nn.Upsample` directly for the upsampling module.
+# - Implement `forward` to wire everything together
+#
+# You'll also need to **copy your helper methods from Task 6A** into the placeholders inside the class — `__init__` and `forward` both call them. Make sure your Task 6A tests pass before you copy them over!
+
+# %% [markdown] tags=[]
+# <div class="alert alert-block alert-info">
+#     <h4>Task 6B: U-Net Implementation</h4>
+#     <ol>
+#         <li>Copy your <code style="color: black">compute_fmaps_encoder</code> and <code style="color: black">compute_fmaps_decoder</code> implementations from Task 6A into the placeholders below.</li>
```
Comment (AlbertDominguez) [inline solution.py]: Instead of having to make students copy code, if we didn't have the UNet stub above, couldn't we just call the function above inside the method?
something like

```python
class UNet(torch.nn.Module):
    def compute_fmaps_encoder(self, level: int):
        return compute_fmaps_encoder(level, self.in_channels, self.num_fmaps, self.fmap_inc_factor) 
```

File: solution.py [hunk 3]
```diff
@@ -661,44 +729,100 @@ def __init__(
...
+        Output (tuple[int, int]): The number of input and output feature maps
+            of the decoder convolutional pass in the given level.
+        """
+        # TASK 6A.2: Implement this function.
+        # Hint: `self.compute_fmaps_encoder` is useful here.
+        ...  # YOUR CODE HERE
+
+
+# %% tags=["solution"]
+class UNet(torch.nn.Module):
```
Comment (AlbertDominguez) [inline solution.py]: I'm not so sure about this stub as it then forces people to actually copy-paste code there below. What would you think if isntead of this, students had to build these functions
```
def compute_fmaps_encoder(level: int, in_channels: int, num_fmaps: int, fmap_inc_factor: int) 
def compute_fmaps_decoder(level: int, in_channels: int, num_fmaps: int, fmap_inc_factor: int) 
```
and then call them directly in the class by propagating the attributes?

**Review by AlbertDominguez — CHANGES_REQUESTED** [review]

File: solution.py (hunk 2 above)
Comment (TeunHuijben) [reply solution.py]: Good idea, implemented!

File: solution.py (hunk 3 above)
Comment (TeunHuijben) [reply solution.py]: Good idea, implemented!
